components/registry-credential/pkg/ecr/updater.go

```diff
@@ -0,0 +1,229 @@
+
+	awsConfig, err := newAWSConfig(region, accessKey, secretKey, "")
+	if err != nil {
+		log.WithError(err).Fatal("unable to new aws config")
+	}
+
+	// Get an authorization token from ECR
+	var (
+		authorizationToken string
+		expiresAt          time.Time
+		endpoint           string
+	)
+	if private {
+		ecrClient := ecr.NewFromConfig(awsConfig)
+		result, err := ecrClient.GetAuthorizationToken(context.TODO(), &ecr.GetAuthorizationTokenInput{})
+		if err != nil {
+			log.WithError(err).Fatal("unable to get an authorization token with private ECR")
+		}
+		if len(result.AuthorizationData) == 0 {
+			log.Fatal("cannot get the authorization data")
+		}
+
+		authorizationToken = aws.ToString(result.AuthorizationData[0].AuthorizationToken)
+		if authorizationToken == "" {
+			log.Fatal("cannot get the authorization token")
+		}
+
+		endpoint = aws.ToString(result.AuthorizationData[0].ProxyEndpoint)
+		if endpoint == "" {
+			log.Fatal("cannot get proxy endpoint")
+		}
+
+		expiresAt = aws.ToTime(result.AuthorizationData[0].ExpiresAt)
+	} else {
+		ecrClient := ecrPublic.NewFromConfig(awsConfig)
+		result, err := ecrClient.GetAuthorizationToken(context.TODO(), &ecrPublic.GetAuthorizationTokenInput{})
+		if err != nil {
+			log.WithError(err).Fatal("unable to get an authorization token from public ECR")
+		}
+		if result.AuthorizationData == nil {
+			log.Fatal("cannot get the authorization data")
+		}
+
+		authorizationToken = aws.ToString(result.AuthorizationData.AuthorizationToken)
+		if authorizationToken == "" {
+			log.Fatal("cannot get the authorization token")
+		}
+
+		expiresAt = aws.ToTime(result.AuthorizationData.ExpiresAt)
+		endpoint = "public.ecr.aws"
+	}
+
+	secretToUpdate, err := getSecret(client, cfg.Namespace, cfg.SecretToUpdate)
+	if err != nil {
+		if !k8serr.IsNotFound(err) {
+			log.WithError(err).Fatalf("cannot find the secret to update %s/%s", cfg.SecretToUpdate, cfg.Namespace)
+		}
+
+		secretToCreate := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cfg.SecretToUpdate,
+				Namespace: cfg.Namespace,
+				Labels: map[string]string{
+					"app":       "gitpod",
+					"component": "registry-credential",
+				},
+			},
+			Type: corev1.SecretTypeDockerConfigJson,
+			StringData: map[string]string{
+				".dockerconfigjson": "{}",
+			},
+		}
+
+		secretToUpdate, err = createSecret(client, cfg.Namespace, secretToCreate)
+		if err != nil {
+			log.WithError(err).Fatalf("cannot create the secret %s/%s", cfg.SecretToUpdate, cfg.Namespace)
+		}
+	}
+
+	err = updateSecretFromToken(client, cfg.Namespace, secretToUpdate, authorizationToken, expiresAt, endpoint)
+	if err != nil {
+		log.WithError(err).Fatalf("Unable to update secret")
+	}
+
+	log.Infof("Secret %s/%s updated with new ECR credentials", cfg.SecretToUpdate, cfg.Namespace)
+}
+
+func newAWSConfig(region, accessKeyId, secretAccessKey, session string) (aws.Config, error) {
```

Comment:
This will need to be changed, for example, to configure AWS given the service account associated with this new component.